src/backend.py

```diff
@@ -231,180 +249,231 @@ def hmac_sha256(key, message):
+
+        # confirm that the challenge is valid
+        if client_challenge not in challenges:
+            return deny_access("Invalid challenge")
 
-        # Retrieve stored password (the key)
+        # confirm that the challenge has not expired
+        if (time() - challenges[client_challenge]) > CHALLENGE_EXPIRATION_SECONDS:
+            del challenges[client_challenge]
+            return deny_access("Challenge expired")
+
+        # Retrieve stored password
         credentials = ds_read_record("configuration", 0)
-        stored_password = credentials["Gpassword"].encode('utf-8')
+        stored_password = credentials["Gpassword"].encode("utf-8")
 
         # Construct the message string
         path = request.path
         query_string = request.query_string or ""
         body_str = request.raw_data or ""
-        message_str = current_challenge + path + query_string + body_str
-        message_bytes = message_str.encode('utf-8')
+        message_str = client_challenge + path + query_string + body_str
+        message_bytes = message_str.encode("utf-8")
 
         # Compute expected HMAC
         expected_digest = hmac_sha256(stored_password, message_bytes)
         expected_hmac = hexlify(expected_digest).decode()
 
+        # If authentication is successful
         if client_hmac == expected_hmac:
-            # Authentication successful
-            current_challenge = None
-            challenge_timestamp = None
+            # remove the challenge from the challenges dict
+            del challenges[client_challenge]
             return handler(request, *args, **kwargs)
         else:
-            # Authentication failed
-            print("Authentication failed")
-            print(f"Expected HMAC: {expected_hmac}")
-            print(f"Client HMAC: {client_hmac}")
-            print(message_bytes)
-            return json_dumps({"error": "Authentication failed"}), 401, 'application/json'
+            return deny_access("Bad Credentials")
 
     return auth_wrapper
 
 
-
 @add_route("/api/auth/challenge")
 def get_challenge(request):
-    global current_challenge, challenge_timestamp
+    global challenges
+
+    # remove expired challenges
+    for challenge, timestamp in list(challenges.items()):
+        if (time() - timestamp) > CHALLENGE_EXPIRATION_SECONDS:
+            del challenges[challenge]
+
+    # make sure there are no more than 10 challenges
+    if len(challenges) > 10:
+        return json_dumps({"error": "Too many challenges"}), 429
+
     # Generate a random nonce (challenge)
-    current_challenge = random_hex(64)
+    new_challenge = random_hex(64)
     challenge_timestamp = time()
+
+    # add the challenge to the challenges dict
+    challenges[new_challenge] = challenge_timestamp
+
     # Return the nonce to the client
-    return json_dumps({"challenge": current_challenge}), 200
+    return json_dumps({"challenge": new_challenge}), 200
+
```

Comment:
Subtle but important change. This enables multiple valid challenges to exist at any instant, this means that two authenticated requests can come in simultaneously without invalidating each other.

also interesting note; I calculated the chance that we get the same challenge twice (which would theoretically allow an attacker to replay past requests without knowing the password) with 256-bits of reasonably random data the chances of a collision is something like 50% at 2^125 samples that should be more than enough. But I found it very interesting that it seems to be half the bits. That could certainly be wrong, I didn't research the exacts very far. I think any number over 10^10 is plenty, and we're at something like 10^37